diff --git a/supervisor/process.py b/supervisor/process.py
index 5f73e6b..00f5b13 100644
--- a/supervisor/process.py
+++ b/supervisor/process.py
@@ -539,6 +539,7 @@ class Subprocess:
             msg = "stopped: %s (%s)" % (processname, msg)
             self._assertInState(ProcessStates.STOPPING)
             self.change_state(ProcessStates.STOPPED)
+            self.config.options.logger.info(msg)
 
         elif too_quickly:
             # the program did not stay up long enough to make it to RUNNING
@@ -548,6 +549,7 @@ class Subprocess:
             msg = "exited: %s (%s)" % (processname, msg + "; not expected")
             self._assertInState(ProcessStates.STARTING)
             self.change_state(ProcessStates.BACKOFF)
+            self.config.options.logger.warn(msg)
 
         else:
             # this finish was not the result of a stop request, the
@@ -569,13 +571,14 @@ class Subprocess:
                 # expected exit code
                 msg = "exited: %s (%s)" % (processname, msg + "; expected")
                 self.change_state(ProcessStates.EXITED, expected=True)
+                self.config.options.logger.info(msg)
             else:
                 # unexpected exit code
                 self.spawnerr = 'Bad exit code %s' % es
                 msg = "exited: %s (%s)" % (processname, msg + "; not expected")
                 self.change_state(ProcessStates.EXITED, expected=False)
+                self.config.options.logger.warn(msg)
 
-        self.config.options.logger.info(msg)
 
         self.pid = 0
         self.config.options.close_parent_pipes(self.pipes)
