diff --git a/relay/dhcrelay.c b/relay/dhcrelay.c
index 4de5621..24bc6e8 100644
--- a/relay/dhcrelay.c
+++ b/relay/dhcrelay.c
@@ -257,6 +257,8 @@ usage(const char *sfmt, const char *sarg) {
 		  isc_file_basename(progname));
 }
 
+void dhcp_relay_usr1_signal_handler(int s);
+
 int 
 main(int argc, char **argv) {
 	isc_result_t status;
@@ -754,6 +756,8 @@ main(int argc, char **argv) {
 	signal(SIGINT, dhcp_signal_handler);   /* control-c */
 	signal(SIGTERM, dhcp_signal_handler);  /* kill */
 #endif
+	/* Using SIGUSR1(user defined signal) to toggle the log level */
+	signal(SIGUSR1, dhcp_relay_usr1_signal_handler);
 
 	/* Start dispatching packets and timeouts... */
 	dispatch();
@@ -2348,3 +2352,20 @@ static void write_stats_to_file(struct interface_info *ip)
       }
    }
 }
+
+void dhcp_relay_usr1_signal_handler(int s)
+{
+	static isc_boolean_t dbg_lvl = ISC_FALSE;
+	if (dbg_lvl == ISC_FALSE)
+	{
+		log_info("Received USR1 signal. Setting log level to debug.");
+		setlogmask(LOG_UPTO(LOG_DEBUG));
+		dbg_lvl = ISC_TRUE;
+	}
+	else
+	{
+		log_info("Received USR1 signal. Resetting log level.");
+		setlogmask(LOG_UPTO(LOG_INFO));
+		dbg_lvl = ISC_FALSE;
+	}
+}
